File: Cpp/027.cpp
#include<iostream>
#define MAX 10000000
using namespace std;

bool prime[MAX];

void init()
{
    prime[1]=false;
    prime[2]=true;
    for(int i=3; i<=MAX; i++)
        if(i%2==0) prime[i]=false;
        else prime[i]=true;
}

void precal()
{
    for(int i=3; i*i<=MAX; i+=2)
    {
        if(prime[i])
        for(int j=i*i; j<=MAX; j+=i)
        {
            prime[j]=false;
        }
    }
}

int calc(int i,int a,int b)
{
    return i*i+a*i+b;
}

int main()
{
    init();
    precal();
    int ans=0,prod;
    for(int a=-1000;a<=1000;a++)
    {
        for(int b=-1000;b<=1000;b++)
        {
            int cnt=0;
            for(int i=0;;i++)
            {
                if (prime[calc(i,a,b)]) cnt++;
                else break;
            }
            if(cnt>ans)
            {
                ans=cnt;
                prod=a*b;
            }
        }
    }
    cout<<prod;
}

File: Ruby/027.rb
MAX = 1_000_000
require_relative 'euler'

sieve

# n in [0, ]
# a in [-999, 999]
# b in primes

max = 0
ans = {}
i = 2
loop do
  if $prime[i]
    b = i
    break if b > 1000
    (-999).upto(999) do |a|
      cnt = 0
      n = 0
      loop do
        if $prime[n * n + a * n + b]
          cnt += 1
        else
          break
        end
        n += 1
      end
      if cnt > max
        max = cnt
        ans = {val: max, a: a, b: b}
      end
    end
  end
  i += 1
end

puts ans
puts ans[:a] * ans[:b]

# Code Golfing Solution:
#
# require 'mathn'
# puts (-999..999).to_a.product((-999..999).to_a).map { |a, b|
#   [(0..100).take_while { |n| (n**2 + a*n + b).prime? }.count, a * b]
# }.max[1]